src/enzyme_ad/jax/Passes/CanonicalizeFor.cpp

```diff
@@ -908,6 +908,120 @@ struct WhileToForHelper {
   }
 };
 
+//This works fine for doWhile with no iter_args case only as of now
+//A more proper approach: copy before region to outside for loop, and copy both before and after region to inside for loop ?
+struct MoveDoWhileToFor : public OpRewritePattern<WhileOp> {
+  using OpRewritePattern<WhileOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(WhileOp whileOp,
+                                PatternRewriter &rewriter) const override {
+      auto module = whileOp->getParentOfType<ModuleOp>();
+      // 1. Trace scf.condition to get condition -> condition is arith.cmpi -> From condition get UB.
+      // 2. Second input of arith.cmpi must be yield value for IV.
+      // 3. From yield value of IV, get the step.
+      // 4. From IV init arg in region get LB
+      // 5. Check after region is just yield of IV for correctness.
+      // 6. Copy region of before body into new created for loop as well as before the for loop.
+      // 7. For outside copy replace uses of inducion var with LB.
+      // 8. Create for loop with for (i=LB+step;i<UB;i+=step)
+      // 9. Replace all uses of 
+
+      //Before block analysis
+      Block &beforeBlock = whileOp.getBefore().front();
+      Value IV = beforeBlock.getArgument(0);
+      auto conditionOp = 
+          dyn_cast<ConditionOp>(beforeBlock.getTerminator());
+      Value loopCondition = conditionOp.getCondition();
+      Value updatedIV = conditionOp.getArgs()[0];
+      Value stepSize;
+      if(auto addOp = updatedIV.getDefiningOp<arith::AddIOp>()){
+        if(addOp.getLhs() == IV) {
+          stepSize = addOp.getRhs(); 
+        }
+        else if(addOp.getRhs() == IV) {
+          stepSize = addOp.getLhs(); 
+        }
+        //Expect atleast one of the updatedIV fields to be IV
+        else 
+          return failure();
+      }
+      Value conditionValue = conditionOp.getCondition();
+      Value lowerBound = whileOp.getOperand(0);
+      
+
+      Value upperBound;
+      if (auto cmpOp = conditionValue.getDefiningOp<arith::CmpIOp>()) {
+        arith::CmpIPredicate predicate = cmpOp.getPredicate();
```

Comment:
Do we need to check the predicate to see if it the loop has a positive step or negative step?